### PR TITLE
OpenSSL model changes arising from ESDK coverage work

### PR DESCRIPTION
**Pull request by tegansb**
*Description of changes:*
Uses the IMPLIES macro
Move all struct and enum definitions from evp_override.c to evp.h header
Adds openssl stubs for _is_valid and _free functions used to improve coverage 
Adds util headers (bn_utils, evp_utils, ec_utils) for helper functions not actually in openssl but useful for CBMC proofs 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: include/bn_utils.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (feliperodri) [inline include/bn_utils.h:2]: Remove all copyrights year. Even better, add the short version instead.
```c
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */
```

File: include/ec_utils.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * permissions and limitations under the License.
+ */
+
+#ifndef EC_UTILS_H
+#define EC_UTILS_H
+
+#include <stdbool.h>
+#include <stdlib.h>
+
+#include <openssl/ec.h>
```
Comment (feliperodri) [inline include/ec_utils.h:22]: Lexicographical order.

File: include/ec_utils.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+size_t max_encryption_size();
+
+/* This function initializes a fixed nondeterministic value meant to represent the maximum possible amount of decrypted
+ * data to be written to the output buffer (see EVP_PKEY_encrypt for an example of its use) */
+void initialize_max_decryption_size();
+
+size_t max_decryption_size();
+
+void write_unconstrained_data(unsigned char *out, size_t len);
```
Comment (feliperodri) [inline include/ec_utils.h:52]: Organize all functions in lexicographical order.

File: include/ec_utils.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+void initialize_max_signature_size();
+
+size_t max_signature_size();
+
+void initialize_max_derivation_size();
+
+size_t max_derivation_size();
+
+/* This function initializes a fixed nondeterministic value meant to represent the maximum possible amount of encrypted
+ * data to be written to the output buffer (see EVP_PKEY_decrypt for an example of its use) */
```
Comment (feliperodri) [inline include/ec_utils.h:41]:
```c
/**
 * Initializes a fixed nondeterministic value meant to represent the maximum
 * possible amount of encrypted data to be written to the output buffer
 * (see EVP_PKEY_decrypt for an example of its use).
 */
```

File: include/ec_utils.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+void initialize_max_encryption_size();
+
+size_t max_encryption_size();
+
+/* This function initializes a fixed nondeterministic value meant to represent the maximum possible amount of decrypted
+ * data to be written to the output buffer (see EVP_PKEY_encrypt for an example of its use) */
```
Comment (feliperodri) [inline include/ec_utils.h:47]:
```c
/**
 * Initializes a fixed nondeterministic value meant to represent the maximum
 * possible amount of decrypted data to be written to the output buffer
 * (see EVP_PKEY_encrypt for an example of its use).
 */
```

File: include/evp_utils.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+bool evp_md_ctx_is_initialized(EVP_MD_CTX *ctx);
+
+size_t evp_md_ctx_get_digest_size(EVP_MD_CTX *ctx);
+
+EVP_PKEY *evp_md_ctx_get0_evp_pkey(EVP_MD_CTX *ctx);
+
+void evp_md_ctx_set0_evp_pkey(EVP_MD_CTX *ctx, EVP_PKEY *pkey);
+
+void evp_md_ctx_shallow_free(EVP_MD_CTX *ctx);
```
Comment (feliperodri) [inline include/evp_utils.h:43]: Lexicographical order.
```c
bool evp_md_ctx_is_initialized(EVP_MD_CTX *ctx);
bool evp_md_ctx_is_valid(EVP_MD_CTX *ctx);
size_t evp_md_ctx_get_digest_size(EVP_MD_CTX *ctx);
EVP_PKEY *evp_md_ctx_get0_evp_pkey(EVP_MD_CTX *ctx);
EVP_MD_CTX *evp_md_ctx_nondet_alloc();
void evp_md_ctx_set0_evp_pkey(EVP_MD_CTX *ctx, EVP_PKEY *pkey);
void evp_md_ctx_shallow_free(EVP_MD_CTX *ctx);
bool evp_pkey_is_valid(EVP_PKEY *pkey);
int evp_pkey_get_reference_count(EVP_PKEY *pkey);
EVP_PKEY *evp_pkey_nondet_alloc();
void evp_pkey_set0_ec_key(EVP_PKEY *pkey, EC_KEY *ec);
void evp_pkey_unconditional_free(EVP_PKEY *pkey);
```

File: include/openssl/evp.h [hunk 7]
```diff
@@ -17,14 +17,63 @@
...
+enum evp_aes { EVP_AES_128_GCM, EVP_AES_192_GCM, EVP_AES_256_GCM };
+enum evp_sha { EVP_SHA256, EVP_SHA384, EVP_SHA512 };
+
+/* Abstraction of the EVP_PKEY struct. */
+struct evp_pkey_st {
+    int references;
+    EC_KEY *ec_key;
+};
+
+/* Abstraction of the EVP_PKEY_CTX struct */
```
Comment (feliperodri) [inline include/openssl/evp.h:37]:
```c
/* Abstraction of the EVP_PKEY_CTX struct. */
```

File: include/openssl/evp.h [hunk 8]
```diff
@@ -17,14 +17,63 @@
...
+struct evp_pkey_ctx_st {
+    bool is_initialized_for_signing;
+    bool is_initialized_for_derivation;
+    bool is_initialized_for_encryption;
+    bool is_initialized_for_decryption;
+    int rsa_pad;
+    EVP_PKEY *pkey;
+};
+
+/* Abstraction of the EVP_CIPHER struct */
```
Comment (feliperodri) [inline include/openssl/evp.h:47]:
```c
/* Abstraction of the EVP_CIPHER struct. */
```

File: include/openssl/evp.h [hunk 9]
```diff
@@ -17,14 +17,63 @@
...
+struct evp_cipher_st {
+    enum evp_aes from;
+    size_t block_size;
+};
+
+/* Abstraction of the EVP_CIPHER_CTX struct */
```
Comment (feliperodri) [inline include/openssl/evp.h:53]:
```c
/* Abstraction of the EVP_CIPHER_CTX struct. */
```

File: include/openssl/evp.h [hunk 10]
```diff
@@ -17,14 +17,63 @@
...
+    EVP_CIPHER *cipher;
+    int encrypt;
+    int iv_len;           // default: DEFAULT_IV_LEN.
+    bool iv_set;          // boolean marks if iv has been set. Default:false.
+    bool padding;         // boolean marks if padding is enabled. Default:true.
+    bool data_processed;  // boolean marks if has encrypt/decrypt final has been called. Default:false.
+    int data_remaining;   // how much is left to be encrypted/decrypted. Default: 0.
+};
+
+/* Abstraction of the EVP_MD struct */
```
Comment (feliperodri) [inline include/openssl/evp.h:64]:
```c
/* Abstraction of the EVP_MD struct. */
```

File: stubs/evp_md_ctx_is_valid_no_pkey_stub.c [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+/*
+ * Checks whether EVP_MD_CTX is a valid object.
+ * Use this stub when we are certain there is no pkey
+ * associated with the digest context.
+ */
+bool evp_md_ctx_is_valid(EVP_MD_CTX *ctx) {
+    assert(ctx->pkey == NULL);
+    return ctx && ctx->is_initialized && ctx->digest_size <= EVP_MAX_MD_SIZE;
+}
```
Comment (feliperodri) [inline stubs/evp_md_ctx_is_valid_no_pkey_stub.c:26]: If `ctx` is NULL, there is a null-pointer dereference in line 24.
```c
/*
 * Checks whether EVP_MD_CTX is a valid object.
 */
bool evp_md_ctx_is_valid(EVP_MD_CTX *ctx) {
    if (!ctx) return false;
    /* We must be sure there is no pkey associated with the digest context. */
    assert(ctx->pkey == NULL);
    return ctx->is_initialized && ctx->digest_size <= EVP_MAX_MD_SIZE;
}
```

File: stubs/evp_pkey_is_valid_no_ec_key_stub.c [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <openssl/evp.h>
+
+/**
+ * Helper function for CBMC proofs: checks if EVP_PKEY is valid.
+ * Use this stub when we are *certain* there is no ec_key associated with the key.
+ */
+bool evp_pkey_is_valid(EVP_PKEY *pkey) {
+    return pkey && (pkey->references > 0) && (pkey->ec_key == NULL);
+}
```
Comment (feliperodri) [inline stubs/evp_pkey_is_valid_no_ec_key_stub.c:24]: If we need to be **certain** there is no `ec_key` associated with the key, don't we need an assertion? If we don't, then we need to update the previous stubs, they must have similar behaviors.
```c
/**
 * Checks if EVP_PKEY is valid.
 */
bool evp_pkey_is_valid(EVP_PKEY *pkey) {
    if (!pkey) return false;
    /* We must be sure there is no ec_key associated with the key. */
    assert(pkey->ec_key == NULL);
    return (pkey->references > 0);
}
```

File: include/bn_utils.h (hunk 1 above)
Comment (feliperodri) [reply include/bn_utils.h:2]: You can just update the date for now and replace all copyrights with the smaller one in a separate PR.

**Review by feliperodri — CHANGES_REQUESTED** [review]
Overall looks good, only a few extra changes.

File: include/bn_utils.h (hunk 1 above)
Comment (tegansb) [reply include/bn_utils.h:2]: I've gone ahead and used the smaller copyright for all the new files I added! Then the rest can be done in a separate PR.

File: include/ec_utils.h (hunk 2 above)
Comment (tegansb) [reply include/ec_utils.h:22]: Done!

File: include/ec_utils.h (hunk 4 above)
Comment (tegansb) [reply include/ec_utils.h:41]: Done

File: include/ec_utils.h (hunk 5 above)
Comment (tegansb) [reply include/ec_utils.h:47]: Done!

File: include/ec_utils.h (hunk 3 above)
Comment (tegansb) [reply include/ec_utils.h:52]: Done!

File: include/evp_utils.h (hunk 6 above)
Comment (tegansb) [reply include/evp_utils.h]: Done!

File: include/openssl/evp.h (hunk 7 above)
Comment (tegansb) [reply include/openssl/evp.h:37]: Done!

File: include/openssl/evp.h (hunk 8 above)
Comment (tegansb) [reply include/openssl/evp.h:47]: Done!

File: include/openssl/evp.h (hunk 9 above)
Comment (tegansb) [reply include/openssl/evp.h:53]: Done!

File: include/openssl/evp.h (hunk 10 above)
Comment (tegansb) [reply include/openssl/evp.h:64]: Done!

File: stubs/evp_md_ctx_is_valid_no_pkey_stub.c (hunk 11 above)
Comment (tegansb) [reply stubs/evp_md_ctx_is_valid_no_pkey_stub.c:16]: Good catch! Updated.

File: stubs/evp_pkey_is_valid_no_ec_key_stub.c (hunk 12 above)
Comment (tegansb) [reply stubs/evp_pkey_is_valid_no_ec_key_stub.c]: Yes, you're right, we should have the assertion here. Updated.

**Review by feliperodri — APPROVED** [review]
LGTM.

**Pull request merged**